src/commands/registries/azure/DockerSiteCreateStep.ts

```diff
@@ -4,37 +4,78 @@
         const creatingNewApp: string = localize('vscode-docker.commands.registries.azure.deployImage.creatingWebApp', 'Creating web app "{0}"...', context.newSiteName);
         ext.outputChannel.appendLine(creatingNewApp);
         progress.report({ message: creatingNewApp });
+        const siteConfig = await this.getSiteConfig(context);
+        const location: AzExtLocation = await LocationListStep.getLocation(context);
+        const locationName: string = nonNullProp(location, 'name');
 
         const armAppService = await import('@azure/arm-appservice');
         const client: WebSiteManagementClient = createAzureClient(context, armAppService.WebSiteManagementClient);
-        context.site = await client.webApps.createOrUpdate(nonNullValueAndProp(context.resourceGroup, 'name'), nonNullProp(context, 'newSiteName'), {
+        const siteEnvelope: Site = {
             name: context.newSiteName,
-            kind: 'app,linux',
-            location: nonNullValueAndProp(context.location, 'name'),
+            location: locationName,
             serverFarmId: nonNullValueAndProp(context.plan, 'id'),
-            siteConfig: this.siteConfig,
-            identity: {
+            siteConfig: siteConfig
+        };
+
+        if (context.customLocation) {
+            // deploying to Azure Arc
+            siteEnvelope.kind = 'app,linux,kubernetes,container';
+        }
+        else {
+            siteEnvelope.identity = {
                 type: 'SystemAssigned'
-            },
-        });
+            };
+        }
+
+        context.site = await client.webApps.createOrUpdate(nonNullValueAndProp(context.resourceGroup, 'name'), nonNullProp(context, 'newSiteName'), siteEnvelope);
+    }
+
+    private async getSiteConfig(context: IAppServiceWizardContext): Promise<WebSiteManagementModels.SiteConfig> {
+        // Temporary workaround until Arc adds support for managed identity, so use usename and password instead.
+        // When customLocation is set, then user is deploying to Arc.
+        const registryTreeItem: RegistryTreeItemBase = this.node.parent.parent;
+        if (registryTreeItem instanceof AzureRegistryTreeItem && context.customLocation) {
+            const appSettings: WebSiteManagementModels.NameValuePair[] = [];
+            const cred = await registryTreeItem.tryGetAdminCredentials();
+            if (!cred) {
+                throw new Error(localize('vscode-docker.commands.registries.azure.dockersitecreatestep.notAdminEnabled', 'Azure App service deployment on Azure Arc only supports running images from Azure Container Registries with admin enabled'));
+            } else {
+                appSettings.push({ name: "DOCKER_REGISTRY_SERVER_URL", value: registryTreeItem.baseUrl });
+                appSettings.push({ name: "DOCKER_REGISTRY_SERVER_USERNAME", value: cred.username });
+                appSettings.push({ name: "DOCKER_REGISTRY_SERVER_PASSWORD", value: nonNullProp(cred, 'passwords')[0].value });

```

Comment:
Does `nonNullProp()` guarantee that there's at least one item in the property array, or just that the property itself is non-null?

Reply:
`notnullProp()` guarantees only the value it not null. The assumption is that if a password is of type array. This was the existing implementation (https://github.com/microsoft/vscode-docker/commit/83f016a589eb247fd2f5d3e3e49096ee73adc8b2#diff-4816881f8a5c146974926a452f283ecd1c0e52099fa13899c7b192ac290e366c)